lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -0,0 +1,1538 @@
+  if (!module) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "Skipping circuit because main module is no `firrtl.module`");
+    return success();
+  }
+  buildDomains(module, InstancePath{}, Value{}, instGraph);
+
+  // Report any domain conflicts among the modules.
+  bool anyFailed = false;
+  for (auto &it : domains) {
+    auto module = cast<FModuleOp>(it.first);
+    auto &domainConflicts = it.second;
+    if (domainConflicts.size() <= 1)
+      continue;
+
+    anyFailed = true;
+    SmallDenseSet<Value> printedDomainResets;
+    auto diag = module.emitError("module '")
+                << module.getName()
+                << "' instantiated in different reset domains";
+    for (auto &it : domainConflicts) {
+      auto &domain = it.first;
+      InstancePathRef path = it.second;
+      auto inst = path.back();
+      auto loc = path.empty() ? module.getLoc() : inst.getLoc();
+      auto &note = diag.attachNote(loc);
+
+      // Describe the instance itself.
+      if (path.empty())
+        note << "Root instance";
+      else {
+        note << "Instance '";
+        llvm::interleave(
+            path, [&](InstanceOp inst) { note << inst.name(); },
+            [&]() { note << "/"; });
+        note << "'";
+      }
+
+      // Describe the reset domain the instance is in.
+      note << " is in";
+      if (domain.reset) {
+        auto nameAndModule = getResetNameAndModule(domain.reset);
+        note << " reset domain rooted at '" << nameAndModule.first.getValue()
+             << "' of module '" << nameAndModule.second.getName() << "'";
+
+        // Show where the domain reset is declared (once per reset).
+        if (printedDomainResets.insert(domain.reset).second) {
+          diag.attachNote(domain.reset.getLoc())
+              << "Reset domain '" << nameAndModule.first.getValue()
+              << "' of module '" << nameAndModule.second.getName()
+              << "' declared here:";
+        }
+      } else
+        note << " no reset domain";
+    }
+  }
+  return failure(anyFailed);
+}
+
+void InferResetsPass::buildDomains(FModuleOp module,
+                                   const InstancePath &instPath,
+                                   Value parentReset, InstanceGraph &instGraph,
+                                   unsigned indent) {
+  LLVM_DEBUG(llvm::dbgs().indent(indent * 2)
+             << "Visiting " << getTail(instPath) << " (" << module.getName()
+             << ")\n");
+
+  // Assemble the domain for this module.
+  ResetDomain domain{.reset = parentReset};
+  auto it = annotatedResets.find(module);
+  if (it != annotatedResets.end()) {
+    domain.isTop = true;
+    domain.reset = it->second;
+  }
+
+  // Associate the domain with this module. If the module already has an
+  // associated domain, it must be identical. Otherwise we'll have to report the
+  // conflicting domains to the user.
+  auto &entries = domains[module];
+  if (entries.empty() || llvm::all_of(entries, [&](const auto &entry) {
```

Comment:
Is the `entries.empty() || ` check redundant?  Can it be removed?

Reply:
Good catch!